esp-hal/src/i2c/master/mod.rs

```diff
@@ -403,13 +403,27 @@ pub struct Config {
 
     /// I2C SCL timeout period.
     timeout: BusTimeout,
+
+    /// Sets the threshold value for the unchanged period of the SCL_FSM state.
+    /// Must not exceed 23. Measurement unit: i2c_sclk.
```

Comment:
This would be better as a named type, then, than validates the value in a const constructor. IIRC it's not done for BusTimeout because the max value depends on the clock source, but you can double check me on that if you'd like.

Reply:
A named type makes sense but I'm not sure about what you have in mind with the const constructor

Reply:
Some abomination like this that would fail in compile time, but force you to fallibly convert in runtime:

```rust
struct Timeout(u8);

impl Timeout {
    pub const fn new_const<const VAL: u8>() -> Self {
        assert!(...)
        ...
    }
}

impl TryFrom<...> for Timeout { ... }
```

Reply:
I thought about s.th. like that but the only way to get the assertion at compile time for me is when I assign it to a const - otherwise it's a panic at runtime

Reply:
You're right it has to be something like:

```rust
    pub const fn new_const<const VAL: u8>() -> Self {
        const { assert(VAL < 24, "Timeout value must be at most 23") }
        ...
    }
```

Reply:
I think something like that made the compiler hate me

Reply:
I wonder if it's better to just silently clamp the value to the max support value, rather than treat it like an error.